p2p/service/gossip_listener.go

```diff
@@ -78,8 +78,41 @@ func (l *Listener) Stop() {
 	l.wg.Wait()
 }
 
-func (l *Listener) listenToGossip(ctx context.Context, dataHandler GossipDataHandler, gossipChannel chan GossipMessage, stop chan struct{}) {
-	l.WithContext(ctx).Info("start listening")
+func (l *Listener) listenToGossip(ctx context.Context, dataHandler GossipDataHandler, gossipChannel chan GossipMessage, stop chan struct{}, channel string) {
+	l.WithContext(ctx).With().Info("start listening to gossip", log.String("protocol", channel))
+
+	// fill the channel with tokens to limit number of concurrent routines
+	tokenChan := make(chan struct{}, l.config.MaxGossipRoutines)
+	for i := 0; i < l.config.MaxGossipRoutines; i++ {
+		tokenChan <- struct{}{}
+	}
+
+	handleMsg := func(ctx context.Context, data GossipMessage) {
+		// get a token to create a new channel
+		l.WithContext(ctx).With().Info("waiting for available slot for gossip handler",
+			log.Int("available_slots", len(tokenChan)),
+			log.Int("total_slots", cap(tokenChan)))
+		if len(tokenChan) == 0 {
+			l.WithContext(ctx).Error("no available slots for gossip handler, blocking")
+		}
+		<-tokenChan
+
+		l.WithContext(ctx).With().Info("got gossip message, forwarding to data handler",
+			log.String("protocol", channel),
+			log.Int("queue_length", len(gossipChannel)))
+		if !l.syncer.ListenToGossip() {
+			// not accepting data
+			l.WithContext(ctx).Info("not currently listening to gossip, dropping message")
+			return
```

Comment:
this looks like token leaking. syncer.ListenToGossip() could become true in the next layer.
i think this check is redundant. it has been checked right before you call handleMsg()

Reply:
it's being checked again here because we may have blocked while waiting for an available token. what do you mean by "token leaking"?

Reply:
i must have misread the code. my bad.